### ISSUE TITLE
Splitting dist/ from git

### ISSUE THREAD

**Issue by himdel** [ManageIQ/ui-components#81]
There's been a few discussion around the topic of not having `dist/` in git, to prevent useless merge conflicts, etc.

Seems like there's 2 options - separate repo for the built package, or gitignore dist/ and git add --force the files. Either relies on travis updating dist/ when doing a successful master build.

Creating the issue so we can have 1 place for the discussion on how to do that...

Cc @mtho11 , @martinpovolny , @karelhala , @skateman

**Comment by skateman** [comment]
I tried the `git add --force` and it simply overwrites the `.gitignore` so after the file is force-added into the repo, the line in `.gitignore` gets ignored ... i.e. it's useless.

IMHO both solutions are terrible, *binaries* should not be present in any git repository. The best would be to drop bower and push to npm using travis, but I know I want too much :stuck_out_tongue_winking_eye:

**Comment by skateman** [comment]
However, I have a completely primitive (and maybe the same terrible) idea: let's drop the binaries and explicitly call  `npm install` and `node_modules/webpack/bin/webpack.js -p` from the `bin/setup` available in `manageiq-ui-classic`. :trollface:  Then we can clean it up when we introduce webpacker.

**Comment by himdel** [comment]
Ok, I guess that settles it :) We need a separate repo if we want to support bower.

Or we can drop bower support, and then we can just npm publish the binaries, yup. (Although that means we can no longer use the master branch as a dependency.)

**Comment by skateman** [comment]
Why not? Can't we create a dummy auto-incremented -dev version after each merge into master? Let's call it unstable branch, or the `sid` :wink:

**Comment by himdel** [comment]
That works too :)

**Comment by skateman** [comment]
Actually, we can do this with bower too, right?

**Comment by himdel** [comment]
Yup, that's true, bower just needs a tag, doesn't strictly need a branch or anything else.

**Comment by skateman** [comment]
So we can create a `dev` tag, a [travis deploy hook](https://github.com/ManageIQ/font-fabulous/blob/master/.travis.yml#L15) with a [script](https://github.com/ManageIQ/font-fabulous/blob/master/ci/gh-pages.sh) and an access token set as a secret ENV var in Travis for **miq-bot** from @Fryguy.

**Comment by Fryguy** [comment]
> drop bower support

:+1:

**Comment by Fryguy** [comment]
What exactly is the issue you are trying to solve though?  Is it just having a place to publish the builds that's not git?  A repo feels like overkill...what about upload to manageiq.org or some other download location?  I believe that could even be done as part of a travis deploy step on merge to master if you need the latest and greatest somewhere.

**Comment by himdel** [comment]
@Fryguy it needs to be on github if we're to support bower. But we can probably drop it now that ui-classic *can* use npm packages (just need to finish the PRs to appliance, docker, etc..)

But the issue we're trying to solve is that by having the generated dist/ files in git, you get a conflict on every PR once any other PR is merged.

**Comment by karelhala** [comment]
I like @skateman's idea of publishing dist under some dummy tag or branch.

**Comment by skateman** [comment]
@karelhala working on it

**Comment by mtho11** [comment]
I like @skateman idea as well. With a preference toward an actual branch over the tag.

**Comment by mtho11** [comment]
Also, can the task of creating the binary be done with an `npm` task so it can be done manually via cli if needed?

**Comment by himdel** [comment]
@mtho11 `yarn run build` ;)

**Comment by mtho11** [comment]
@himdel perfect then! 👍

**Comment by himdel** [comment]
Sure it can :) To retrieve a particular version, it needs a tag, but right now we're depending on `    "manageiq-ui-components": "master"` and that's just the branch name.

**Comment by skateman** [comment]
Just because dealing with branches is a bit more complicated as I need to sort out conflicts :disappointed:

**Comment by himdel** [comment]
Maybe you don't .. if we say that that particular branch is always autogenerated, and if we run the script only when the tests passed *and* we're on master, then there's no confclicts, right? Or at least, we can always just pull the branch, overwrite all the files with the current version, and push --force it.

**Comment by himdel** [comment]
(At least, I see no reason to generate `dist/` from any other source branch but master.)

**Comment by skateman** [comment]
![nothingtodohere](https://user-images.githubusercontent.com/649130/28131455-35543818-673a-11e7-9652-fd464b5c3fea.gif)

**Comment by mtho11** [comment]
Just for completeness: https://bower.io/docs/api/#install

**Comment by Fryguy** [comment]
Keeping the discussion here, in #83 it seems the approach is to create a dummy commit that is tagged with `dev`, off of master, so something like this...

```text
(master)
    o - - o (dev)
    |
    |
    o - - o
    |
    |
    o - - o
    |
    |
```

My concerns is that the PR seems to force push the same tag name for each commit, which seems bad IMO.  You generally *never* want to force push anything that's public with git.  Additionally, you lose the builds from the previous commits, because the tag is lost and thus the commit is not reachable.  So, when someone clones they can't get the dist values at old points in time.

A `tag` is a static reference, whereas a `branch` is actually a moving reference, which seems close to what is trying to be accomplished here, so perhaps it's better to a branch instead?  It can possibly be an [orphan branch](https://git-scm.com/docs/git-checkout/1.7.3.1#git-checkout---orphan), or if only the dist directory is needed can be a completely separate branch like most people do with gh-pages branches.  So I'm picturing something like this:

```text
(master) (dev)
    o     o 
    |     |
    |     |
    o     o
    |     |
    |     |
    o     o
    |
    |
```

Then, when you create the new commit on dev, write the SHA of the corresponding commit on master into the commit message.  Then you can correlate every build with every master commit quite easily.

@skateman Thoughts?

**Comment by Fryguy** [comment]
> > Can bower deal with a branch?
> 
> Sure it can :) To retrieve a particular version, it needs a tag, but right now we're depending on "manageiq-ui-components": "master" and that's just the branch name.

Can you give it a SHA instead of a label ref, if needed?

**Comment by himdel** [comment]
Agreed it should be a branch, not a tag. Pretty sure bower caches tags...

**Comment by skateman** [comment]
`yarn run build` is failing with:
```
error Command failed with exit code 2.
```

**Comment by himdel** [comment]
@skateman And how does that help us without the actual error?

**Comment by skateman** [comment]
Well, I can't see anything else...

**Comment by himdel** [comment]
Ah, ok then :) works for me... maybe try running `webpack -p` manually..
